onnxruntime/core/providers/rocm/rocm_profiler.cc

```diff
@@ -0,0 +1,239 @@
+
+  char buff[BSIZE];
+
+  for (auto &item : d->rows_) {
+    std::initializer_list<std::pair<std::string, std::string>> args = {{"op_name", ""}};
+    addEventRecord(item, profiling_start, args, event_map);
+  }
+
+  for (auto &item : d->mallocRows_) {
+    snprintf(buff, BSIZE, "%p", item.ptr);
+    const std::string arg_ptr{buff};
+
+    std::initializer_list<std::pair<std::string, std::string>> args = {{"op_name", ""},
+                                            {"ptr", arg_ptr},
+                                            {"size", std::to_string(item.size)}
+                                            };
+    addEventRecord(item, profiling_start, args, event_map);
+  }
+
+  for (auto &item : d->copyRows_) {
+    snprintf(buff, BSIZE, "%p", item.stream);
+    const std::string arg_stream{buff};
+    snprintf(buff, BSIZE, "%p", item.src);
+    const std::string arg_src{buff};
+    snprintf(buff, BSIZE, "%p", item.dst);
+    const std::string arg_dst{buff};
+
+    std::initializer_list<std::pair<std::string, std::string>> args = {{"op_name", ""},
+                                            {"stream", arg_stream},
+                                            {"src", arg_src},
+                                            {"dst", arg_dst},
+                                            {"size", std::to_string(item.size)},
+                                            {"kind", std::to_string(item.kind)},
+                                            };
+    addEventRecord(item, profiling_start, args, event_map);
+    copyLaunches[item.id] = &item;
+  }
+
+  for (auto &item : d->kernelRows_) {
+    snprintf(buff, BSIZE, "%p", item.stream);
+    const std::string arg_stream{buff};
+    if (item.functionAddr)
+        snprintf(buff, BSIZE, "%s", demangle(hipKernelNameRefByPtr(item.functionAddr, item.stream)).c_str());
+    else if (item.function)
+        snprintf(buff, BSIZE, "%s", demangle(hipKernelNameRef(item.function)).c_str());
+    else
+        snprintf(buff, BSIZE, " ");
+    const std::string arg_kernel{buff};
+
+    std::initializer_list<std::pair<std::string, std::string>> args = {{"op_name", ""},
+                                            {"stream", arg_stream},
+                                            {"kernel", arg_kernel},
+                                            {"grid_x", std::to_string(item.gridX)},
+                                            {"grid_y", std::to_string(item.gridY)},
+                                            {"grid_z", std::to_string(item.gridZ)},
+                                            {"block_x", std::to_string(item.workgroupX)},
+                                            {"block_y", std::to_string(item.workgroupY)},
+                                            {"block_z", std::to_string(item.workgroupZ)},
+                                            };
+    addEventRecord(item, profiling_start, args, event_map);
+    kernelLaunches[item.id] = &item;
+  }
+
+  // Async Ops - e.g. "Kernel"
```

Comment:
What does this comment mean?

Reply:
Due to how roctracer works, I am collecting "api" timestamps myself.  That is, libroctracer is giving me an inline callback (from the user's thread) both pre and post the actual call into the hip function.  I am recording cpu timestamps and args and sticking them in my internal structures (e.g. d->kernelRows_) until profiling has ended.

It makes no sense to have an "inline callback" for a gpu operation.  Those are logged and timestamped by libroctracer.  When a page of those is full I get a callback (from one of libroctracer's threads) where I copy them off into d->gpuTraceBuffers_.  Don't even peak at them while profiling is on.

So everything above line 120 comment is an api call, hereafter a onnxruntime::profiling::API_EVENT.  That is a new event type I added to capture cpu operations related to the cuda/rocm apis.  E.g. hipmalloc, hipLaunchKernel.  The cupti profiler does not capture these.  But should.

The "async ops" below line 120 are the classic onnxruntime::profiling::KERNEL_EVENT events.  Things that executed on the gpu.  These are what the cupti profiler currently reports.

Yeah, comment could be better.

Reply:
Got you, thanks! Can you please elaborate the comment a bit?